### ISSUE TITLE
Interface for class based middleware?

### ISSUE THREAD

**Issue by praswicaksono** [relayphp/Relay.Relay#6]
I think is better to make interface for class based middleware instead just write instruction in docs to implement __invoke

**Comment by pmjones** [comment]
I thought about that. Then I realized that If one _has_ to implement an interface, that means closures are no longer allowed (e.g., `function ($request, $response)`). I'd like to allow for that use case, which means adding an interface is not that useful.

**Comment by praswicaksono** [comment]
As I stated in title it will be applied on 'Class Based' middleware only

for example we ensure that class resolved is implements interface

``` php
    protected function resolve($entry)
    {
        if (! $entry) {
            // the default callable when the queue is empty
            return function (
                Request $request,
                Response $response,
                callable $next
            ) {
                return $response;
            };
        }
        if (! $this->resolver) {
            return $entry;
        }
        $middleware = call_user_func($this->resolver, $entry);
        if (! $middleware instanceof MiddlewareInterface) {
             throw new \Exception("Middleware must implement MiddlewareInterface");
         }

         return $middleware;
    }
```

So it will accept closure and MiddlewareInterface

**Comment by pmjones** [comment]
Again, I considered this, but the resolver might well be a registry of closures. I get where you're coming from on the use of interfaces, but there's a strong existing bias in middleware-land that allows for closures, which just can't be type hinted.

**Comment by mindplay-dk** [comment]
> there's a strong existing bias in middleware-land that allows for closures, which just can't be type hinted.

Closure can be type-hinted with php-doc, for IDE support, and the interface would help formalize the implementation of middleware classes, as well as serving as documentation.

I currently have this:

``` PHP
use Psr\Http\Message\ServerRequestInterface as Request;
use Psr\Http\Message\ResponseInterface as Response;

/**
 * This interface fortifies the middleware interface signature used by Relay.
 */
interface MiddlewareInterface
{
    /**
     * @param Request  $request
     * @param Response $response
     * @param callable $next
     *
     * @return Response
     */
    public function __invoke(Request $request, Response $response, callable $next = null);
}
```

And when I type-hint against it, I do this:

``` PHP
/**
 * @param callable|MiddlewareInterface $middleware
 */
public function run(callable $middleware)
{
    // ...
}
```

In other words, just don't type-hint statically with MiddlewareInterface, but do type-hint with annotations for IDE support, inspections, auto-completion, etc. - for example, a list of middlware is:

``` PHP
/**
 * @var (callable|MiddlewareInterface)[] $middlewares
 */
```

There's no harm in having the interface as guidance for implementation, just don't use it in static (argument) type-hints, use it in (`@var`, `@param`) annotations only, it'll never even load, will provide great IDE support, and won't get in the way of anything.

PR? :-)

**Comment by pmjones** [comment]
@mindplay-dk PR! Also, a readme PR to incorporate your docblock+typehint suggestion, which I had not thought of. Thanks!

**Comment by mindplay-dk** [comment]
I'll see if I can get around to this tomorrow :-)

**Comment by mindplay-dk** [comment]
I'm implementing this, and wondering if `$request` in this interface should be type-hinted as PSR `RequestInterface` or `ServerRequestInterface`?

You type-hinted `$request` in `FakeMiddleware::__invoke()` as `RequestInterface`, which made me wonder if that was deliberate.

**Comment by mindplay-dk** [comment]
Also, sadly, for the record, the latest version of Php Storm no longer parses php-doc type-hints like `(callable|MiddlewareInterface)[]` correctly - this used to work fine, I don't know what happened. In terms of IDE support (in Storm anyway) that means type-hints for arrays will receive no better or worse support than before. On the upside, it will serve as documentation, and single instance type-hints such as `callable|MiddlewareInterface` do work as expected.

**Comment by pmjones** [comment]
Closed per #13 . @mindplay-dk I see that PhpStorm is not acting the way you'd like; if you have further PRs along this line, I can merge them as well. My apologies for the delay.
